sql: prevent time-related precision during upgrade from 19.2 -> 20.1

otan opened the pull request:
The intention is to backport this if we think this change is important.

Release note (sql change): We previously allowed mixed type 20.1/19.2
upgrades to add precision types. However, the 19.2 would disrespect the
precision component and we documented that. To make this cleaner, we
instead only allow precision to be specified in time-related types when
the version upgrade is complete.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47425)
<!-- Reviewable:end -->

otan commented:
bors r=rohany

jordanlewis commented on pkg/sql/create_table.go:
```diff
@@ -119,12 +119,27 @@ var minimumTypeUsageVersions = map[types.Family]clusterversion.VersionKey{
+func isTypeSupportedInVersion(v clusterversion.ClusterVersion, t *types.T) (bool, error) {
+	switch t.Family() {
+	case types.TimeFamily, types.TimestampFamily, types.TimestampTZFamily:
+		if t.Precision() != 6 && !v.IsActive(clusterversion.VersionTimePrecision) {
```
Also TimeTZFamily?

otan replied:
that one is already gated below!

otan replied:
(but might as well anyway)

otan commented:
bors r-

andreimatei commented:
nit: don't say things like "this PR" in the release notes. When release notes are published, there's no "this". Please see guidance in the commit msg template and on the wiki.

otan commented:
gotcha

otan committed:
sql: prevent time-related precision during upgrade from 19.2 -> 20.1

The intention is to backport this if we think this change is important.

Release note (sql change): We previously allowed mixed type 20.1/19.2
upgrades to add precision types. However, the 19.2 would disrespect the
precision component and we documented that. To make this cleaner, we
instead only allow precision to be specified in time-related types when
the version upgrade is complete.

otan commented:
bors r=rohany
